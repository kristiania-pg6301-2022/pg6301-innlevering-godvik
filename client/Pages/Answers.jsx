import { Link, Route, Routes } from "react-router-dom";
import React from "react";

export function Answers() {
  return (
    <div>
      <Routes>
        <Route path="wrong" element={<h1>WRONG ANSWER</h1>} />
        <Route path="correct" element={<h1>CORRECT ANSWER</h1>} />
      </Routes>
      <button>
        <Link to="/quiz">New question</Link>
      </button>
      <button>
        <Link to="/">Show score</Link>
      </button>
    </div>
  );
}
