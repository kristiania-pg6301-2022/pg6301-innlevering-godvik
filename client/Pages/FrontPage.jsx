import { Link } from "react-router-dom";
import React from "react";

export function FrontPage({ score, answeredQuestions }) {
  return (
    <div>
      <h1>Welcome to the QUIZ!!!</h1>
      <div>
        You have answered {score} of {answeredQuestions} correctly
      </div>
      <Link to={"/quiz"}>Take a quiz</Link>
    </div>
  );
}
