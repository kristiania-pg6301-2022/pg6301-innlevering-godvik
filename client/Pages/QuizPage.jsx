import React from "react";
import { useLoader } from "../utils/hooks/useLoader";
import { fetchJSON } from "../utils/http";
import { Link } from "react-router-dom";
import { useNavigate } from "react-router";

export function QuizPage({ setScore, setAnsweredQuestions }) {
  const { loading, error, data, reload } = useLoader(
    async () => await fetchJSON("/api/question")
  );
  const navigate = useNavigate();
  const question = data;

  if (loading) {
    return <h1>Loading....</h1>;
  }

  if (error) {
    return (
      <div style={{ border: "1px solid red", background: "Pink" }}>
        An error occurred: {error.toString()}
      </div>
    );
  }

  function quizHandler(answer) {
    const { id } = question;
    fetch(`/api/question/`, {
      method: "post",
      headers: {
        "content-type": "application/json",
      },
      body: JSON.stringify({ answer, id }),
    })
      .then((res) => res.json())
      .then((res) => {
        if (res.result) {
          setScore((score) => score + 1);
          setAnsweredQuestions((answeredQuestions) => answeredQuestions + 1);
          navigate("/answers/correct");
        } else {
          navigate("/answers/wrong");
          setAnsweredQuestions((answeredQuestions) => answeredQuestions + 1);
        }
      });
  }

  return (
    <div>
      <h1>{question.question}</h1>
      {Object.keys(question.answers)
        .filter((a) => question.answers[a])
        .map((answer) => (
          <div key={answer}>
            <button onClick={() => quizHandler(answer)}>
              {question.answers[answer]}
            </button>
          </div>
        ))}
      <Link to="/">
        <button>Back</button>
      </Link>
    </div>
  );
}
