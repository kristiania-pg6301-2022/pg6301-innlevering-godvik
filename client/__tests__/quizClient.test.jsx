import React from "react";
import ReactDOM from "react-dom";
import { FrontPage } from "../Pages/FrontPage";
import { MemoryRouter } from "react-router-dom";
import { QuizPage } from "../Pages/QuizPage";
import { Answers } from "../Pages/Answers";
import pretty from "pretty";

describe("QuizApp client testing", () => {
  it("Shows answer status", () => {
    const element = document.createElement("div");
    ReactDOM.render(
      <MemoryRouter>
        <FrontPage initialEntries={["/"]} score={3} answeredQuestions={10} />
      </MemoryRouter>,
      element
    );
    expect(element.querySelector("[data-testid=status]").textContent).toEqual(
      "You have answered 3 of 10 correctly"
    );
    expect(pretty(element.innerHTML)).toMatchSnapshot();
  });

  it("snapshot test for wrong answers", () => {
    const div = document.createElement("div");
    ReactDOM.render(
      <MemoryRouter>
        <Answers initialEntries={["/wrong"]} />
      </MemoryRouter>,
      div
    );
    expect(pretty(div.innerHTML)).toMatchSnapshot();
  });

  it("snapshot test for right answers", () => {
    const div = document.createElement("div");
    ReactDOM.render(
      <MemoryRouter>
        <Answers initialEntries={["/correct"]} />
      </MemoryRouter>,
      div
    );
    expect(pretty(div.innerHTML)).toMatchSnapshot();
  });
});
