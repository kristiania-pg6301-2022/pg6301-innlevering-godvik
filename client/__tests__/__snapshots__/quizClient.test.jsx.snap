// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`QuizApp client testing Shows answer status 1`] = `
"<div>
  <h1>Welcome to the QUIZ!!!</h1>
  <div data-testid=\\"status\\">You have answered 3 of 10 correctly</div><a href=\\"/quiz\\">Take a quiz</a>
</div>"
`;

exports[`QuizApp client testing snapshot test for right answers 1`] = `"<div><button><a href=\\"/quiz\\">New question</a></button><button><a href=\\"/\\">Show score</a></button></div>"`;

exports[`QuizApp client testing snapshot test for wrong answers 1`] = `"<div><button><a href=\\"/quiz\\">New question</a></button><button><a href=\\"/\\">Show score</a></button></div>"`;
